### PR TITLE
Use new Dictionary AlternateLookup to avoid allocation in DictionaryJumpTable

### PR DESCRIPTION
**Pull request by andrewjsaid**
Use new Dictionary AlternateLookup to avoid allocation in DictionaryJumpTable

## Description

Simple change which removes the `substring` allocation in DictionaryJumpTable using the `GetAlternateLookup<ReadOnlySpan<char>>` feature of .NET 9.

With this change all JumpTable implementations can accept `ReadOnlySpan<char>` as a parameter instead of `string, PathSegment`. I'd be happy to include that change if it's wanted.

### CODE REVIEW

**Comment by BrennanConroy** [comment]
Awesome, was looking forward to aspnetcore adopting the new alternate lookup feature.

Could you run the micro-benchmarks before and after this change?
https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/perf/Microbenchmarks/Matching/JumpTableMultipleEntryBenchmark.cs

File: src/Http/Routing/perf/Microbenchmarks/Matching/JumpTableMultipleEntryBenchmark.cs [hunk 1]
```diff
@@ -27,7 +27,7 @@ public void Setup()
 
         for (var i = 0; i < _strings.Length; i++)
         {
-            _segments[i] = new PathSegment(0, _strings[i].Length);
+            _segments[i] = new PathSegment(1, _strings[i].Length - 2);
```
Comment (BrennanConroy) [inline src/Http/Routing/perf/Microbenchmarks/Matching/JumpTableMultipleEntryBenchmark.cs:30]: Shouldn't this be `- 1`?

**Review by BrennanConroy — APPROVED** [review]

File: src/Http/Routing/perf/Microbenchmarks/Matching/JumpTableMultipleEntryBenchmark.cs (hunk 1 above)
Comment (andrewjsaid) [reply src/Http/Routing/perf/Microbenchmarks/Matching/JumpTableMultipleEntryBenchmark.cs:30]: I attempted to trim one from the start and one from the end but I don't suppose it matters too much as long as it's a substring.

**Comment by andrewjsaid** [comment]
I've created https://github.com/dotnet/runtime/pull/108732 which if accepted/merged should further improve the performance of the AlternateLookup used in this PR.

**Comment by BrennanConroy** [comment]
/azp run

Was waiting for the runtime PR so we could see the full perf gains. But since it's taking time to get a review we can just merge this since it's still goodness.

**Comment by davidfowl** [comment]
@BrennanConroy did we ever re-run the benchmarks?

**Pull request merged**